UM-USElab/gradecraft-development
# Copy files

### Status
READY

### Description

These changes are an extension of the work performed in #3794, which involves resolving copy lookups. The changes in this PR allow the following to be copied as well:

1. Badge icon
2. Badge files
3. Assignment media
4. Assignment files
5. Challenge media
6. Challenge files
7. Course syllabus

The process is that the files themselves will be copied on S3, rather than just having the file in the copied model reference the file in the original.

Notes: Two notable changes were made to the `ModelCopier` class. One involves switching the order of operations for performing lookups and performing overrides and the other involves making the original model readonly so as to reduce any unintended side-effects.

**Future changes**

- Need to address polymorphic unlock conditions, which are not currently able to be copied via #3794 
- Need to access files always using presigned urls so that we can remove public ACLs

**Todo**

- Ensure that remote files in S3 are deleted on model destroy, if we want to be mindful of storage usage
- Ensure that when copying a specific assignment, the assignment files are actually being replicated in S3 rather than simply being linked to the original file.

### Deploy Notes
**As an interim solution, public ACLs on the applicable buckets in S3 must be created to ensure that badge icons are visible post-copy**

1. Badge icon
2. Assignment media
3. Challenge files

See current bucket policy for gradecraft-staging for additional details.

### Gem dependencies
- UPDATED `acts_as_list` to latest (0.9.10), to allow suppressing of unwanted behavior that happens behind the scenes on model updates.

### Migrations
NO

### Steps to Test or Reproduce

- May need to ensure that there are no errant `BadgeFile`s in production, since this adds a new validation to ensure that a `BadgeFile` must have a file.

### Impacted Areas in Application
Course copy

======================
Closes #3770
